provider/aws: Adding some other simple S3 Bucket Object (Optional) Inputs

stack72 opened the pull request:
Added the following to the S3 Bucket Object:
*  cache_control
*  content_disposition
*  content_encoding
*  content_language
*  content_type

Tests pass as expected:
```
make testacc TEST=./builtin/providers/aws TESTARGS='-run=S3BucketObject' 2>~/tf.log
go generate ./...
TF_ACC=1 go test ./builtin/providers/aws -v -run=S3BucketObject -timeout 90m
=== RUN TestAccAWSS3BucketObject_basic
--- PASS: TestAccAWSS3BucketObject_basic (4.64s)
=== RUN TestAccAWSS3BucketObject_withContentCharacteristics
--- PASS: TestAccAWSS3BucketObject_withContentCharacteristics (3.45s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	8.109s
```

stack72 commented:
@catsby while you are on a merging spree, this one should be a pretty simple addition :) It just adds support for some more pieces of the S3 puzzle

catsby commented on website/source/docs/providers/aws/r/s3_bucket_object.html.markdown:
```diff
@@ -29,6 +29,11 @@ The following arguments are supported:
+* `content_disposition` - (Optional) Specifies presentational information for the object.
+* `content_encoding` - (Optional) Specifies what content encodings have been applied to the object and thus what decoding mechanisms must be applied to obtain the media-type referenced by the Content-Type header field.
+* `content_language` - (Optional) The language the content is in.
+* `content_type` - (Optional) A standard MIME type describing the format of the object data.
```
Can we throw in a few examples of these? e.g. for `cache_control`, two examples would be `no-cache`, or `max-age=3000`? Is the `max-age=3000` even valid? Some guidance would help people (like myself...) on how to use it correctly

catsby commented on builtin/providers/aws/resource_aws_s3_bucket_object.go:
```diff
@@ -61,9 +91,14 @@ func resourceAwsS3BucketObjectPut(d *schema.ResourceData, meta interface{}) erro
+			ContentDisposition: aws.String(contentDisposition),
+			ContentEncoding:    aws.String(encoding),
+			ContentLanguage:    aws.String(contentLanguage),
+			ContentType:        aws.String(contentType),
```
Because these fields are optional, `d.Get("content_encoding").(string)` will return an empty string `""` if it's not included in the configuration file. We'll then pass that on in the `PutObjectIntput` struct, so `CacheControl` will equal `""`. In most of the AWS API calls, this is not correct, and can introduce unexpected or even undefined behavior. 

We should check if the values are `""` and conditionally add them to the struct

stack72 commented:
@catsby thanks for reviewing this. I have added links to the correct sections of w3c docs for some of the more complex documentation (e.g. cache_control and content_disposition) and then added some samples for content_type and content_language

Also changed the PutObject input to check for empty strings before adding to the struct

stack72 commented:
@catsby ok, I have added some explanations here and changed to support the empty string checks

FYI, master had removed the Update method and 'Forced New' on everything. So I was getting this error:

```
--- FAIL: TestProvider-2 (0.00s)
	provider_test.go:24: err: aws_s3_bucket_object: No Update defined, must set ForceNew on: []string{"cache_control", "content_disposition", "content_encoding", "content_language", "content_type"}
```

By making the same changes to my code, the world was a better place :)

I *think* this may be ready for you now

catsby commented on builtin/providers/aws/resource_aws_s3_bucket_object.go:
```diff
@@ -26,6 +25,36 @@ func resourceAwsS3BucketObject() *schema.Resource {
+			"content_type": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
```
Seems like at least this one needs to be `computed` as well, I'm getting a plan loop with this config: 

```hcl
provider "aws" {
  region = "us-west-2"
}

resource "aws_s3_bucket" "object_bucket" {
	bucket = "tf-object-test-bucket-1234"
}

resource "aws_s3_bucket_object" "object" {
	bucket = "${aws_s3_bucket.object_bucket.bucket}"
	key = "test-key"
	source = "terraform.tfstate"
}
```

```console
~ aws_s3_bucket_object.object
    content_type: "binary/octet-stream" => ""
```

stack72 replied:
@catsby can a param be both Optional and Computed?

stack72 replied:
actually I take that back, by default, an S3 object has a default Content-Type of 'binary/octet-stream' so I will update to do that

catsby commented:
One final bit and we're good to go

stack72 commented:
@catsby we are good with this now :)

catsby commented:
Thanks @stack72 !
